[Clang] Allow malloc for functions returning structs

a-nogikh opened the pull request:
The malloc attribute restricts the possible function signatures to the ones returning a pointer, which is not the case for some non-standard allocation functions variants. For example, P0901R11 proposed ::operator new overloads that return a return_size_t result - a struct that contains a pointer to the allocated memory as well as the actual size of the allocated memory. Another example is __size_returning_new.

Relax the mentioned restriction and allow the "malloc" attribute to be applied not just to the functions returning pointer types, but also to the functions returning records whose first member is a pointer (which would be assumed to point to the allocated memory). This is the case for return_size_t as well as std::span, should it be returned from such an annotated function.

As the previous restriction was dictated by the need to set the noalias attribute, adjust the code generation logic to only assign the attribute when the returned value type is a pointer type.

In future commits, codegen can be improved to recognize the noalias-ness of the pointer returned inside a span-like struct.

This change also helps unlock the alloc token instrumentation for such non-standard allocation functions:
https://clang.llvm.org/docs/AllocToken.html#instrumenting-non-standard-allocation-functions

a-nogikh committed:
[Clang] Allow malloc and alloc_size attributes for functions returning structs

These attributes restrict the possible function signatures to the ones
returning a pointer, which is not the case for some non-standard
allocation functions variants. For example, P0901R11 proposed ::operator new
overloads that return a return_size_t result - a struct that contains
a pointer to the allocated memory as well as the actual size of the
allocated memory. Another example is __size_returning_new.

Relax the mentioned restriction and allow "malloc" and "alloc_size"
attributes to be applied not just to the functions returning pointer
types, but also to the functions returning records whose first member is
a pointer (which would be assumed to point to the allocated memory).
This is the case for return_size_t as well as std::span, should it be
returned from such an annotated function.

As the previous restriction was dictated by the need to set the noalias
attribute, adjust the code generation logic to only assign the attribute
when the returned value type is a pointer type.

In future commits, codegen can be improved to recognize the
noalias-ness of the pointer returned inside a span-like struct.

This change also helps unlock the alloc token instrumentation for such
non-standard allocation functions:
https://clang.llvm.org/docs/AllocToken.html#instrumenting-non-standard-allocation-functions

efriedma-quic reviewed (commented):
Needs release note.  Needs documentation update for the attribute.

Since this attribute was originally defined by gcc, we should consult with gcc devs to see if they have a position on this.  It would be a proble if gcc chooses some different behavior here.

AaronBallman reviewed (commented):
> Since this attribute was originally defined by gcc, we should consult with gcc devs to see if they have a position on this. It would be a proble if gcc chooses some different behavior here.

+1 to this request

pinskia commented:
> This LGTM, but I'd like to hear from others more familiar with the history of these attributes.
> 
> In particular, we'd be diverging from GCC's behaviour, so that users would now need to do:
> 
> ```
> #ifdef __clang__
> #define __attribute_malloc_span  __attribute__((malloc))
> #define __attribute_alloc_size_span(...)  __attribute__((alloc_size(__VA_ARGS__)))
> #else
> #define __attribute_malloc_span  
> #define __attribute_alloc_size_span(...) 
> #endif 
> ```

Actually it is worse you would need to do something like:
```
#ifdef __clang__
  #if __clang_major__ >= 22
    #define have_attribute_malloc_span
  #endif
#endif

#ifdef have_attribute_malloc_span
   #define __attribute_malloc_span  __attribute__((malloc))
   #define __attribute_alloc_size_span(...)  __attribute__((alloc_size(__VA_ARGS__)))
 #else
   #define __attribute_malloc_span  
    #define __attribute_alloc_size_span(...) 
 #endif
```
if you want to support older clang.

a-nogikh committed:
fixup: move isSpanLikeType() to SemaDeclAttr.cpp

Also leave a comment about possible false positives.

a-nogikh committed:
fixup: update the documentation

Mention the change in the release notes and extend the attribute
descriptions.

a-nogikh committed:
fixup: drop alloc_size attribute changes

Extending the range of functions alloc_size applies to makes no sense
without the __builtin_object_size-side support.

For now, just the malloc attribute changes should be enough.

a-nogikh commented:
Thank you for the feedback!

* I've sent an email to gcc@gcc.gnu.org: https://gcc.gnu.org/pipermail/gcc/2025-November/246931.html
* Dropped the `alloc_size`-related changes from the PR: after some more consideration, they didn't seem really necessary and only complicated the changeset.
* Added a bullet to ReleaseNotes and extended the documentation for the attribute.

***

> Is there a reason why a new attribute name can't be used instead? Like say `malloc_span` and `alloc_size_span`. This should make easier to keep things working when trying with say `__has_attribute` and older clang/gcc and you don't need to hard code the exact version of clang/gcc where the support was added. I know GCC's malloc attribute added an extra argument which is hard to test for except via a feature test via an autoconf like but I feel like that was a mistake [one which we need to live with now] (I didn't think of the issue back when GCC added that support nor I responded to the review back then either).

If adding a new `malloc_span` attribute will simplify the usage / prevent incompatibilities with GCC, that sounds totally reasonable to me, thanks for the suggestion. I can prepare an alternative PR that will do this instead.

AaronBallman commented:
> Is there a reason why a new attribute name can't be used instead? Like say `malloc_span` and `alloc_size_span`. This should make easier to keep things working when trying with say `__has_attribute` and older clang/gcc and you don't need to hard code the exact version of clang/gcc where the support was added. I know GCC's malloc attribute added an extra argument which is hard to test for except via a feature test via an autoconf like but I feel like that was a mistake [one which we need to live with now] (I didn't think of the issue back when GCC added that support nor I responded to the review back then either).

I think that's a reasonable idea; I'd be happy with `malloc_span`

a-nogikh commented:
A pull request with the `malloc_span` implementation: https://github.com/llvm/llvm-project/pull/167010
